### PR TITLE
Sampler rework ugla

### LINKED ISSUE #381: Subclass the UGLA sampler from the new Sampler class

### PR DESCRIPTION
**Pull request by chaozg**
fixed #381 

- We followed a similar strategy as in [NUTS](https://github.com/CUQI-DTU/CUQIpy/blob/main/cuqi/experimental/mcmc/_hmc.py), and the pre-computation work is now done in `_pre_warmup` and `_pre_sample` functions.
- A local target is created within `test_UGLA_regression_sample`, instead of a global target list, to avoid breaking randomness in other unit tests.
- Further work is needed in the future to base UGLA on LinearRTO #410 .

### CODE REVIEW

File: tests/zexperimental/test_mcmc.py [hunk 1]
```diff
@@ -273,6 +300,7 @@ def test_CWMH_regression_warmup(target: cuqi.density.Density):
     cuqi.experimental.mcmc.ULANew(cuqi.testproblem.Deconvolution1D().posterior, scale=0.0001),
     cuqi.experimental.mcmc.MALANew(cuqi.testproblem.Deconvolution1D().posterior, scale=0.0001),
     cuqi.experimental.mcmc.LinearRTONew(cuqi.testproblem.Deconvolution1D().posterior),
+    # cuqi.experimental.mcmc.UGLANew(create_lmrf_prior_target(dim=16)) #TODO: this cause a error in MHNew
```
Comment (nabriis) [inline tests/zexperimental/test_mcmc.py:303]: Please rebase on main, then if needed, update the scale in MHNew in checkpoint_targets to e.g. scale=0.0001. Should fix the issue.

File: tests/zexperimental/test_mcmc.py [hunk 2]
```diff
@@ -239,6 +239,33 @@ def test_RegularizedLinearRTO_regression_warmup(target: cuqi.density.Density):
     sampler_new = cuqi.experimental.mcmc.RegularizedLinearRTONew(target, stepsize=1e-3)
     assert_true_if_warmup_is_equivalent(sampler_old, sampler_new)
 
+def create_lmrf_prior_target(dim=16):
```
Comment (nabriis) [inline tests/zexperimental/test_mcmc.py:242]: After rebasing on main there might be another place to add the sampler. The place related to state.

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+        # If prior location is scalar, repeat it to match dimensions
+        if len(self.prior.location) == 1:
+            self._priorloc = np.repeat(self.prior.location, self.dim)
+        else:
+            self._priorloc = self.prior.location
+
+        # Initial Laplace approx
+        # self._L2 = Lk_fun(self.x0)
+        self._L2 = Lk_fun(np.zeros(self.dim)) #TODO: fix this
```
Comment (nabriis) [inline cuqi/experimental/mcmc/_laplace_approximation.py:105]: I see the issue with the "_precompute" method here. I think we need to follow a design pattern similar to what Amal did in the NUTS sampler. Use a "_pre_sample" and "_pre_warmup" methods. Have these called before sampling to define parameters. See e.g. in https://github.com/CUQI-DTU/CUQIpy/pull/387

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+import scipy as sp
+import numpy as np
+import cuqi
+from cuqi.solver import CGLS
+from cuqi.experimental.mcmc import SamplerNew
+
+class UGLANew(SamplerNew):
```
Comment (nabriis) [inline cuqi/experimental/mcmc/_laplace_approximation.py:7]: Perhaps this was dicussed in a daily update at some point, but I was hoping that LinearRTO and UGLA could use the same base class. If not in this PR, lets at least create an issue for it.

**Review by nabriis — CHANGES_REQUESTED** [review]
Hi @chaozg. Thanks for this PR. Overall looks good. I don't have any major comments, but I attempted to help you in the two problems you listed.
Let me know if the guidance is unclear.

File: tests/zexperimental/test_mcmc.py (hunk 1 above)
Comment (chaozg) [reply tests/zexperimental/test_mcmc.py]: Done as suggested

File: tests/zexperimental/test_mcmc.py (hunk 2 above)
Comment (chaozg) [reply tests/zexperimental/test_mcmc.py:249]: I also added `UGLANew` to `state_history_targets`

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 3 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py]: Done as suggested

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 4 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py:7]: I see your point: UGLA shares a lot of code with RTOs except for the additional approximation step. For now, I create  an issue for it #410.

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 3 above)
Comment (nabriis) [reply cuqi/experimental/mcmc/_laplace_approximation.py]: Looks good to me. Leaving unresolved for next reviewer to see.

**Review by nabriis — APPROVED** [review]
Thanks for dealing with the comments @chaozg. LGTM

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    
+    Samples an approximate posterior where the prior is approximated
+    by a Gaussian distribution. The likelihood must be Gaussian.
+
+    Currently only works for LMRF priors.
+
+    The inner solver is Conjugate Gradient Least Squares (CGLS) solver.
+
+    For more details see: Uribe, Felipe, et al. "A hybrid Gibbs sampler for edge-preserving 
+    tomographic reconstruction with uncertain view angles." arXiv preprint arXiv:2104.06919 (2021).
```
Comment (jakobsj) [inline cuqi/experimental/mcmc/_laplace_approximation.py:18]: Update ref to published version: https://doi.org/10.1137/21M1412268

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    Parameters
+    ----------
+    target : `cuqi.distribution.Posterior`
+        The target posterior distribution to sample.
+
+    initial_point : ndarray
+        Initial parameters. *Optional*
```
Comment (jakobsj) [inline cuqi/experimental/mcmc/_laplace_approximation.py:26]: Specify default value.

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    maxit : int
+        Maximum number of inner iterations for solver when generating one sample.
+
+    tol : float
+        Tolerance for inner solver. Will stop before maxit if the inner solvers convergence check reaches tol.
+
+    beta : float
+        Smoothing parameter for the Gaussian approximation of the Laplace distribution. Larger beta is easier to sample but is a worse approximation.
```
Comment (jakobsj) [inline cuqi/experimental/mcmc/_laplace_approximation.py:35]: Could we help the user by specifying a  range or typical values? Less than 1? 1.0e12?

File: cuqi/experimental/mcmc/_laplace_approximation.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    target : `cuqi.distribution.Posterior`
+        The target posterior distribution to sample.
+
+    initial_point : ndarray
+        Initial parameters. *Optional*
+
+    maxit : int
+        Maximum number of inner iterations for solver when generating one sample.
+
+    tol : float
```
Comment (jakobsj) [inline cuqi/experimental/mcmc/_laplace_approximation.py:31]: Any default, or typical values we could specify to guide the user?

**Review by jakobsj — CHANGES_REQUESTED** [review]
Many thanks @chaozg, it looks good to me and I have only added a few very minor comments.

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 5 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py:18]: Done

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 6 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py:26]: Done

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 8 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py:34]: Done

File: cuqi/experimental/mcmc/_laplace_approximation.py (hunk 7 above)
Comment (chaozg) [reply cuqi/experimental/mcmc/_laplace_approximation.py]: Done after group discussion

**Review by jakobsj — APPROVED** [review]

**Pull request merged**